### ISSUE TITLE
[Bug]: Error in applying recdevs when SRR is type 7 (survivorship)

### ISSUE THREAD

**Issue by Rick-Methot-NOAA** [nmfs-stock-synthesis/stock-synthesis#420]
### Describe the bug

user reported bad Hessian with some parameter settings for SRR == survivorship.
First examination disclosed that the Spawn_Recruit table had similar values for all columns, but only the last column should have had rec_dev applied.

Examination of the SS3 code led to finding that this code sequence in popdyn and forecast:
```
        Recruits = Spawn_Recr(SSB_use, R0_use, SSB_current); // calls to function Spawn_Recr
        apply_recdev(Recruits, R0_use); //  apply recruitment deviation
```
works for all other SRR functions because they only fill exp_rec(y,1) then apply_recdev fills the other columns.
But survivorship is a specialized dev so all columns are filled with the SRR function.  
Therefore, the attempted solution will be to not call apply_recdev() if SRR is using survivorship.

The suggested solution solves the spawn_recr output table.
Hessian is still nondef if Beta != 1.0
Next try setting Beta = 1.0 only for equil calcs and allow the non-zero value for exp(recr).
This works and produces a posdef Hessian. 
But Why????  Leave that for another day.
### To Reproduce

per above

### Expected behavior

correct spawn_recr table and good Hessian

**Comment by iantaylor-NOAA** [comment]
This issue appears when I just changed the west coast dogfish model from SR_surv_Beta fixed at 1.0 to 1.5.

Interestingly, the parameters themselves seem to be estimated with uncertainty and the correlations among them are reported in covar.sso, but the derived quantities all have 0 variance and at the end of estimation it reports
```
Differentiating 8 derived quantities: 5 600 -log(L): 419.302  Spbio: 31963.5 13444                                                                                          1, 2, 3, 4, 5, 6, 7, 8 done (16 m 11 s)
Inverting Hessian (70 variables): 20%, 40%, 60%, 80%, 100% done (0.019s)
Starting standard error calculations...
Warning: Non-positive variance of sdreport variables: 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 
...
637, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648, 649
done (3.4 s)   
```
Model files are in this folder (accessible only to NOAA folks): https://drive.google.com/drive/folders/1uNhSZFfxUSV5Lw8eDXTJpYeSCnLMuLq0?usp=sharing

**Comment by Rick-Methot-NOAA** [comment]
In SRR=7, the recdevs are applied to survival inside the SRR-7 code.  So something abut that does not like beta !=1.

**Comment by Rick-Methot-NOAA** [comment]
No solution yet, but have looped in Johnoel to see if it might be an ADMB issue.
While doing deep dive into performance of this code I noticed that this change seems more correct:
```
      NewRecruits = exp_rec(y, 3) * SRZ_surv;  //  using bias adjusted as base for dev
//      NewRecruits = SSB_curr_adj * SRZ_surv;
 ```
My tests have changed from zero-centered devs to normal devs
Also have eliminated the time-varying growth
neither fixes the sd_variable variance problem

**Comment by Rick-Methot-NOAA** [comment]
next rewrote formulas with pow() to use exp(b*ln(a)) instead.  This had no impact on problem.
Then noticed in forecast-report.sso that the search for Btarget and for MSY are failing and MSY search showing -0 values.  This is highly probably related to the problem finding variance on initial year SSB, which then cascades to cause fail of all other

**Comment by Rick-Methot-NOAA** [comment]
Solution:
The benchmark search explores F levels that can produce negative B_equil for this SRR.  Other SRR have a posfun() to trap for this.  adding the posfun() to this SRR solves the problem.

Additional issues addressed here:
In the estimation of recruitment during the time series, the bias adjustment factor was not being applied for this SRR.  This has been corrected:
`      NewRecruits = exp_rec(y, 3) * SRZ_surv;  //  using bias adjusted as base for dev
//      NewRecruits = SSB_curr_adj * SRZ_surv;`

Finally, this SRR is using SSB_virgin, not SSB_unfished.  This difference matters when there is time-varying growth (per Miller and Brooks Slippery Slope paper).  This issue should be addressed separately and may require adding a new version of SRR_7 with this alternative

**Comment by Rick-Methot-NOAA** [comment]
@iantaylor-NOAA putting this PR back into attention.

**Comment by iantaylor-NOAA** [comment]
The changes in the draft pull request look good. I'm running a model to test it as well.

**Comment by iantaylor-NOAA** [comment]
Just confirming (as @Rick-Methot-NOAA already knew), that a model with non-integer SR_surv_Beta produced variances of derived quantities after this fix now merged in #421.
Also, the stock-recruit plot in r4ss had looked strange before and that is now fixed. Previously the line for the expected recruitment was not smooth and now it is. I had assumed this was an r4ss bug but should have followed up on it which might have led to this issue getting identified earlier.

**Comment by Rick-Methot-NOAA** [comment]
Note that the posfun was added to solve the problem, but the 3rd parameter was left in disabled state.
Will do new issue to re-enable the 3rd parameter.
